src/getty-generator/getty-generator.c

```diff
@@ -25,41 +25,35 @@
 static const char *arg_dest = NULL;
 static bool arg_enabled = true;
 
-static int add_serial_getty(const char *tty) {
-        _cleanup_free_ char *instance = NULL;
+static int add_getty_impl(const char *tty, const char *unit, const char *type) {
         int r;
 
         assert(tty);
+        assert(unit);
+        assert(type);
 
         tty = skip_dev_prefix(tty);
 
-        log_debug("Automatically adding serial getty for /dev/%s.", tty);
-
+        _cleanup_free_ char *instance = NULL;
         r = unit_name_path_escape(tty, &instance);
         if (r < 0)
-                return log_error_errno(r, "Failed to escape tty path: %m");
+                return log_error_errno(r, "Failed to escape tty path %s: %m", tty);
 
-        return generator_add_symlink_full(arg_dest,
-                                          "getty.target", "wants",
-                                          SYSTEM_DATA_UNIT_DIR "/serial-getty@.service", instance);
-}
-
-static int add_container_getty(const char *tty) {
-        _cleanup_free_ char *instance = NULL;
-        int r;
+        _cleanup_free_ char *unit_path = path_join(SYSTEM_DATA_UNIT_DIR, unit);
+        if (!unit_path)
+                return log_oom();
 
-        assert(tty);
-        assert(!path_startswith(tty, "/dev/"));
+        log_debug("Automatically adding %s getty for %s.", type, tty);
 
-        log_debug("Automatically adding container getty for /dev/pts/%s.", tty);
+        return generator_add_symlink_full(arg_dest, "getty.target", "wants", unit_path, instance);
+}
 
-        r = unit_name_path_escape(tty, &instance);
-        if (r < 0)
-                return log_error_errno(r, "Failed to escape tty path: %m");
+static int add_serial_getty(const char *tty) {
+        return add_getty_impl(tty, "serial-getty@.service", "serial");
+}
 
-        return generator_add_symlink_full(arg_dest,
-                                          "getty.target", "wants",
-                                          SYSTEM_DATA_UNIT_DIR "/container-getty@.service", instance);
+static int add_container_getty(const char *tty) {
+        return add_getty_impl(tty, "container-getty@.service", "container");
```

Comment:
Hmm, this doesn't really enable getty.ttys.container to take a pty path, as the unit only accepts a pty name. Please move the validation and skipping of /dev/pts/ prefix into this function instead.